### ISSUE TITLE
V8: Language fallback extraction error when deploying between environments

### ISSUE THREAD

**Issue by HalldorLyngmo** [umbraco/Umbraco.Cloud.Issues#88]
When you have a couple of languages in the backoffice and have configured a fallback that is not English (United States) you will encounter an extraction error when you deploy it to Live

It will give a similar error message to this.
`Cannot save language is with fallback id=3 which is not a valid language id.`

Reproduction
------------
### Bug summary

I have a Dev environment and have added two languages, Icelandic and Danish. Then I add Danish as a fallback for Icelandic. 

When I do that the uda file will have id 3 as the **FallbackLanguageId**. Then I deploy it to Live, but I run into an extraction error. Because it can't save the Icelandic language when there is no language with id 3.

When you deploy the languages without a fallback you can see that they don't get the same id on the next environment. Also when you add the languages manually on the Live you can't guarantee that it will have the same id as they have on Dev.

So this means that you will always have some issue with the fallback if you use anything other than English (United States) which has id 1 as default. If you remove it and add it again it might get something else as an id and become an issue.

The solution could be to use the alias of the fallback language rather than the id of it.

### Specifics
```
EXCEPTION:

Umbraco.Deploy.Exceptions.ProcessArtifactException: Process pass #0 failed for artifact umb://language/is.
   at Umbraco.Deploy.Environments.CurrentEnvironment.ProcessArtifacts(IEnumerable`1 arts, IDeployContext context, CancellationToken token)
   at Umbraco.Deploy.Environments.CurrentEnvironment.ProcessPackageAsync(Guid sessionId, Package package, CancellationToken token)
   at Umbraco.Deploy.Work.WorkItems.DiskReadWorkItem.<Proceed>d__17.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Umbraco.Deploy.Work.WorkItems.DiskReadWorkItem.<ExecuteAsyncSub>d__16.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Umbraco.Deploy.Work.WorkItems.DiskReadWorkItem.<ExecuteAsync>d__15.MoveNext()

INNER:
	System.InvalidOperationException: Cannot save language is with fallback id=3 which is not a valid language id.
	   at Umbraco.Core.Services.Implement.LocalizationService.Save(ILanguage language, Int32 userId)
	   at Umbraco.Deploy.Connectors.ServiceConnectors.LanguageConnector.Process(ArtifactDeployState`2 state, IDeployContext context, Int32 pass)
	   at Umbraco.Deploy.Connectors.ServiceConnectors.ServiceConnectorBase`3.Process(ArtifactDeployState dart, IDeployContext context, Int32 pass)
	   at Umbraco.Deploy.Environments.CurrentEnvironment.ProcessArtifacts(IEnumerable`1 arts, IDeployContext context, CancellationToken token)

EXCEPTION:

Umbraco.Deploy.Exceptions.ProcessArtifactException: Process pass #0 failed for artifact umb://language/is.
   at Umbraco.Deploy.Environments.CurrentEnvironment.ProcessArtifacts(IEnumerable`1 arts, IDeployContext context, CancellationToken token)
   at Umbraco.Deploy.Environments.CurrentEnvironment.ProcessPackageAsync(Guid sessionId, Package package, CancellationToken token)
   at Umbraco.Deploy.Work.WorkItems.DiskReadWorkItem.<Proceed>d__17.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Umbraco.Deploy.Work.WorkItems.DiskReadWorkItem.<ExecuteAsyncSub>d__16.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Umbraco.Deploy.Work.WorkItems.DiskReadWorkItem.<ExecuteAsync>d__15.MoveNext()

INNER:
	System.InvalidOperationException: Cannot save language is with fallback id=3 which is not a valid language id.
	   at Umbraco.Core.Services.Implement.LocalizationService.Save(ILanguage language, Int32 userId)
	   at Umbraco.Deploy.Connectors.ServiceConnectors.LanguageConnector.Process(ArtifactDeployState`2 state, IDeployContext context, Int32 pass)
	   at Umbraco.Deploy.Connectors.ServiceConnectors.ServiceConnectorBase`3.Process(ArtifactDeployState dart, IDeployContext context, Int32 pass)
	   at Umbraco.Deploy.Environments.CurrentEnvironment.ProcessArtifacts(IEnumerable`1 arts, IDeployContext context, CancellationToken token)



[Failed]



EXCEPTION
Umbraco.Deploy.Exceptions.ProcessArtifactException: Process pass #0 failed for artifact umb://language/is. ---> System.InvalidOperationException: Cannot save language is with fallback id=3 which is not a valid language id.
   at Umbraco.Core.Services.Implement.LocalizationService.Save(ILanguage language, Int32 userId)
   at Umbraco.Deploy.Connectors.ServiceConnectors.LanguageConnector.Process(ArtifactDeployState`2 state, IDeployContext context, Int32 pass)
   at Umbraco.Deploy.Connectors.ServiceConnectors.ServiceConnectorBase`3.Process(ArtifactDeployState dart, IDeployContext context, Int32 pass)
   at Umbraco.Deploy.Environments.CurrentEnvironment.ProcessArtifacts(IEnumerable`1 arts, IDeployContext context, CancellationToken token)
   --- End of inner exception stack trace ---
   at Umbraco.Deploy.Environments.CurrentEnvironment.ProcessArtifacts(IEnumerable`1 arts, IDeployContext context, CancellationToken token)
   at Umbraco.Deploy.Environments.CurrentEnvironment.ProcessPackageAsync(Guid sessionId, Package package, CancellationToken token)
   at Umbraco.Deploy.Work.WorkItems.DiskReadWorkItem.<Proceed>d__17.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Umbraco.Deploy.Work.WorkItems.DiskReadWorkItem.<ExecuteAsyncSub>d__16.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Umbraco.Deploy.Work.WorkItems.DiskReadWorkItem.<ExecuteAsync>d__15.MoveNext()

```

### Steps to reproduce
1. Create a project with two environments.
2. Add two extra languages on Dev.
3. Add one of the extra languages as a fallback
4. Deploy them to Live
5. Extraction error - Cannot save language is with fallback id=3 which is not a valid language id.

### Expected result
It should be deploye without an issue and I should see all the languages on Live with the right fallback.

### Actual result
Extraction error and can't deploy it like this

### Workaround
Only adding the fallback on Live can work but it can be overwritten if there are any changes done on Dev.

---
_This item has been added to our backlog [AB#8099](https://dev.azure.com/umbraco/243e7927-03b2-44e2-908f-d4ac7ea5daaa/_workitems/edit/8099)_

**Comment by OlePc** [comment]
You will get the same error by following these steps:

1. Create a project
2. Add two extra languages on Live
3. Set the second language as fallback for the first
3. Delete the default language (en-US)
4. Clone to localhost
5. Fire up the site
6. Notice `Cannot save language [Your language code] with fallback id=3 which is not a valid language id.` error


Can be fixed by changing fallback from 3 to null in the uda file, and switch back after the initial deploy is done. Still annoying though.

**Comment by KasperBHansenAtPentia** [comment]
This is also blocking deployment i my project!

**Comment by thomashdk** [comment]
it looks like the identity value wen from 1 to 5
So my English got Id 1 as on the other projects and my danish language got id 5, but on the other projects it has id 2.

I removed the danish language and runned this query on the table dbo.umbracoLanguage to reset the indentity...
`DBCC CHECKIDENT ('umbracoLanguage', RESEED, 1)  `

Then I created the danish language in the backoffice again and now it has the correct id ( 2 )..

Maybe a little hacky.. but it did the work for me..

**Comment by ronaldbarendse** [comment]
This is caused by the fact the language UDA contains an integer ID for the fallback, which can vary per environment, e.g.:
```json
{
  "IsoCode": "nl",
  "Alias": "nl",
  "IsDefault": false,
  "IsMandatory": false,
  "FallbackLanguageId": 1,
  "Udi": "umb://language/nl",
  "Dependencies": [],
  "Name": "Dutch",
  "__type": "Umbraco.Deploy,Umbraco.Deploy.Artifacts.LanguageArtifact",
  "__version": "3.4.5"
}
```

Umbraco Deploy should convert this ID to the ISO code (e.g. `en`) when exporting the UDA-file and after deploying all languages (adding new and updating existing), update the fallback language in a second stage.

**Comment by ronaldbarendse** [comment]
@nul800sebastiaan This currently completely blocks deployment whenever the language IDs don't match between environments and could even result in wrong fallbacks being used if the IDs are swapped!

One possible workaround might be to completely disable deployment of languages by adding `<add type="language" />` to the [excluded entity types](https://our.umbraco.com/documentation/Umbraco-Cloud/Deployment/Deploy-Settings/#excludedentitytypes) and manually apply the changes between environments.

**Comment by ronaldbarendse** [comment]
Looks like the excluded entity types only work for content deploys, not metadata (the UDI files), so there's currently no workaround 🤯 @nul800sebastiaan Is it possible to increase the priority of this issue, as it's blocking all deployments when a single fallback language is configured?

Because the integer ID in the `FallbackLanguageId` is useless (might not exist in other environments or otherwise reference an incorrect language), the fix can be quite easy: replace this property on `LanguageArtifact` with the ISO code of the fallback language (e.g. `FallbackLanguageIsoCode`) and update the `LanguageConnector` to map this back to the integer ID of the current environment.

I'm not sure if the `LanguageArtifact` requires backwards compatibility, but if that's the case, the `FallbackLanguageId` property could be attributed with `Obsolete` and `JsonIgnore` instead of completely removing it...

**Comment by ronaldbarendse** [comment]
When this gets fixed, it should set the fallback language in a separate step after all languages are deployed, like explained in issue https://github.com/umbraco/Umbraco.Cloud.Issues/issues/188.

**Comment by clausjensen** [comment]
This has been fixed in Deploy 3.5.0

**Comment by ronaldbarendse** [comment]
@clausjensen Thanks, but what's the hold-up though? I see the 3.5.0 release is created almost a month ago (both in the nightly and MyGet repositories) and this fix would finally make it possible to configure fallback languages on Umbraco Cloud projects (as they otherwise can't be restored/deployed, because of the mismatching integer IDs)!
